test/test_view_subarray.jl

```diff
@@ -1,19 +1,217 @@
-function test_view_subarray()
-  snlp = SimpleNLPModel(x -> sum(x.^4), ones(3),
-                        g=x -> 4x.^3,
-                        c=x -> [dot(x, x) - 4.0; sum(x) - 1.0],
-                        J=x -> [2x[1]  2x[2]  2x[3]; ones(1,3)],
-                        H=(x;y=zeros(2),obj_weight=1.0) -> sparse([1, 2, 3], [1, 2, 3],
-                                                                  12obj_weight * x.^2 +
-                                                                  2 * y[1] * ones(3)),
-                        lcon=zeros(2), ucon=zeros(2))
-  adnlp = ADNLPModel(x -> sum(x.^4), ones(3), c=x -> [dot(x, x) - 4.0; sum(x) - 1.0])
-  x = rand(5)
-  nlps = [snlp, adnlp]
-  functions = [obj, grad, hess, cons, jac]
-  for nlp in nlps, foo in functions, I = [1:3, 1:2:5, [2;4;5], [4;1;3]]
-    @test foo(nlp, x[I]) == foo(nlp, @view x[I])
+using Random
+
+function test_view_subarray_nlp(nlp)
+  @testset "Test view subarray of NLPs" begin
+    n, m = nlp.meta.nvar, nlp.meta.ncon
+    N = 2n
+    Vidxs = [1:n, n.+(1:n), 1:2:N, shuffle(1:N)[1:n]]
+    Cidxs = if m > 0
+      N = 2m
+      [1:m, m.+(1:m), 1:2:N, shuffle(1:N)[1:m]]
+    else
+      []
+    end
+
+    # Inputs
+    x  = rand(2n)
+    v  = rand(2n)
+    y  = rand(2m)
+
+    # Outputs
+    g    = zeros(n)
+    g2   = zeros(2n)
+    c    = zeros(m)
+    c2   = zeros(2m)
+    jv   = zeros(m)
+    jv2  = zeros(2m)
+    jty  = zeros(n)
+    jty2 = zeros(2n)
+    hv   = zeros(n)
+    hv2  = zeros(2n)
+
+    for I = Vidxs
+      xv = @view x[I]
+      for foo in (obj, grad, hess)
+        @test foo(nlp, x[I]) ≈ foo(nlp, xv)
+      end
+
+      rows1, cols1, vals1 = hess_coord(nlp, x[I])
+      rows2, cols2, vals2 = hess_coord(nlp, xv)
+      @test rows1 == rows2
+      @test cols1 == cols2
+      @test vals1 ≈ vals2
+
+      if m > 0
+        for foo in (cons, jac)
+          @test foo(nlp, x[I]) ≈ foo(nlp, xv)
+        end
+        rows1, cols1, vals1 = jac_coord(nlp, x[I])
+        rows2, cols2, vals2 = jac_coord(nlp, xv)
+        @test rows1 == rows2
+        @test cols1 == cols2
+        @test vals1 ≈ vals2
+      end
+
+      for J = Cidxs
+        yv = @view y[J]
+        @test hess(nlp, x[I], y=y[J]) ≈ hess(nlp, xv, y=yv)
+        yv = @view y[J]
+        rows1, cols1, vals1 = hess_coord(nlp, x[I], y=y[J])
+        rows2, cols2, vals2 = hess_coord(nlp, xv, y=yv)
+        @test rows1 == rows2
+        @test cols1 == cols2
+        @test vals1 ≈ vals2
+      end
+
+      # Inplace methods can have input and output as view, so 4 possibilities
+      for J = Vidxs
+        gv = @view g2[J]
+        grad!(nlp, x[I],  g)
+        grad!(nlp, x[I], gv); @test g ≈ g2[J]
```

Comment:
Should this compare g to gv instead of g2?

Reply:
`gv` is the view for `g2[J]`, so they should be the same. Do you want me to change?

Reply:
Ah, sorry. No need to change. It looks good.